subsys/bluetooth/gatt_pool.c

```diff
@@ -235,121 +266,188 @@ static void uuid_unregister(struct bt_uuid const *uuid)
+	if (!gp || !gp->svc.attrs || !svc_uuid) {
+		LOG_ERR("Invalid attribute");
+		return -EINVAL;
+	}
+	if (gp->svc.attr_count >= gp->attr_array_size) {
+		LOG_ERR("No space left on given svc");
+		return -ENOSPC;
+	}
 
 	memset(attr, 0, sizeof(*attr));
 
-	uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_primary);
-	uuid_register((struct bt_uuid **) &attr->user_data, svc_uuid);
+	ret = uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_primary);
+	if (ret) {
+		return ret;
+	}
+	ret = uuid_register((struct bt_uuid **) &attr->user_data, svc_uuid);
+	if (ret) {
+		return ret;
+	}
 	attr->perm = BT_GATT_PERM_READ;
 	attr->read = bt_gatt_attr_read_service;
 
-	__ASSERT_NO_MSG(svc->attr_count < CONFIG_BT_GATT_HIDS_ATTR_MAX);
-	svc->attr_count++;
-}
-
-void bt_gatt_pool_svc_put(struct bt_gatt_attr const *attr)
-{
-	__ASSERT_NO_MSG(attr != NULL);
-
-	uuid_unregister(attr->uuid);
-	uuid_unregister(attr->user_data);
+	gp->svc.attr_count++;
+	return 0;
 }
 
-void bt_gatt_pool_chrc_get(struct bt_gatt_service *svc,
-			   struct bt_gatt_chrc const *chrc)
+int bt_gatt_pool_chrc_alloc(struct bt_gatt_pool *gp,
+			    struct bt_gatt_chrc const *chrc)
 {
-	struct bt_gatt_attr *attr = &svc->attrs[svc->attr_count];
+	int ret;
+	struct bt_gatt_attr *attr = &gp->svc.attrs[gp->svc.attr_count];
 	struct bt_uuid      *uuid_gatt_chrc = BT_UUID_GATT_CHRC;
 	struct bt_gatt_chrc *dest_chrc;
 
-	__ASSERT_NO_MSG(svc->attrs != NULL);
+	if (!gp || !gp->svc.attrs || !chrc) {
+		LOG_ERR("Invalid attribute");
+		return -EINVAL;
+	}
+	if (gp->svc.attr_count >= gp->attr_array_size) {
+		LOG_ERR("No space left on given svc");
+		return -ENOSPC;
+	}
 
 	memset(attr, 0, sizeof(*attr));
 
-	uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_chrc);
+	ret = uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_chrc);
+	if (ret) {
+		return ret;
+	}
 	attr->perm = BT_GATT_PERM_READ;
 	attr->read = bt_gatt_attr_read_chrc;
 
 	/* Register user data for characteristic. */
-	chrc_get((struct bt_gatt_chrc **) &attr->user_data);
+	ret = chrc_get((struct bt_gatt_chrc **) &attr->user_data);
+	if (ret) {
+		return ret;
+	}
 	dest_chrc = (struct bt_gatt_chrc *) attr->user_data;
 	memset(dest_chrc, 0, sizeof(*dest_chrc));
 	dest_chrc->properties = chrc->properties;
-	uuid_register((struct bt_uuid **) &dest_chrc->uuid, chrc->uuid);
-
-	__ASSERT_NO_MSG(svc->attr_count < CONFIG_BT_GATT_HIDS_ATTR_MAX);
-	svc->attr_count++;
-}
-
-void bt_gatt_pool_chrc_put(struct bt_gatt_attr const *attr)
-{
-	__ASSERT_NO_MSG(attr != NULL);
+	ret = uuid_register((struct bt_uuid **) &dest_chrc->uuid, chrc->uuid);
+	if (ret) {
+		return ret;
+	}
 
-	uuid_unregister(((struct bt_gatt_chrc *) attr->user_data)->uuid);
-	chrc_release(attr->user_data);
-	uuid_unregister(attr->uuid);
+	gp->svc.attr_count++;
+	return 0;
 }
 
-void bt_gatt_pool_desc_get(struct bt_gatt_service *svc,
-			   struct bt_gatt_attr const *descriptor)
+int bt_gatt_pool_desc_alloc(struct bt_gatt_pool *gp,
```

Comment:
Note: Becouse pool free is automatic now, and it is based on attr uuid field, I would suggest to add here a check if it is not the same like any of the used in predefined types.